### PR TITLE
Add linear extroplation to LinearInterpolation

### LINKED ISSUE #15379: Add linear extrapolation to evaluate points that lie outside given data set domain.
## Reason

LinearInterpolation does not perform extrapolation for points that lie outside given data set domain. For some cases, it would be nice to provide an option to perform extrapolation using two points nearest the endpoint. 

## Design

Add option for extrapolation. By default, that option is set to be false. 

## Impact

New feature added.

### PR DESCRIPTION
**Pull request by jiangwen84**
closes  #15379

### CODE REVIEW

**Comment by jiangwen84** [comment]
@dschwen Could you review this PR? Thanks.

File: framework/src/utils/LinearInterpolation.C [hunk 1]
```diff
@@ -70,10 +87,22 @@ T
...
-    return 0.0;
-  if (x >= _x[_x.size() - 1])
-    return 0.0;
+  if (_extrap)
+  {
+    if (x <= _x[0])
+      return (_y[1] - _y[0]) / (_x[1] - _x[0]);
+    if (x >= _x.back())
+    {
+      return (_y[_x.size() - 2] - _y.back()) / (_x[_x.size() - 2] - _x.back());
```
Comment (dschwen) [inline framework/src/utils/LinearInterpolation.C:96]: shouldn't the first one technically be a `_y.size()` ?

Comment (dschwen) [reply framework/src/utils/LinearInterpolation.C:96]: well, they should be the same

Comment (jiangwen84) [reply framework/src/utils/LinearInterpolation.C:96]: Good catch! Should be _y.size(). Copied pasted wrong one.

**Review by dschwen — DISMISSED** [review]

File: framework/src/utils/LinearInterpolation.C (hunk 1 above)
Comment (jiangwen84) [reply framework/src/utils/LinearInterpolation.C]: Yes, they are the same, but I fix it. Thanks.

File: framework/src/utils/LinearInterpolation.C [hunk 2]
```diff
@@ -70,10 +87,22 @@ T
...
-  if (x >= _x[_x.size() - 1])
-    return 0.0;
+  if (_extrap)
+  {
+    if (x <= _x[0])
+      return (_y[1] - _y[0]) / (_x[1] - _x[0]);
+    if (x >= _x.back())
+    {
+      return (_y[_y.size() - 2] - _y.back()) / (_x[_x.size() - 2] - _x.back());
+    }
```
Comment (dschwen) [inline framework/src/utils/LinearInterpolation.C:97]: no braces around single line statements

**Review by dschwen — DISMISSED** [review]
The style issue can be addressed later if you need this PR now

File: framework/src/utils/LinearInterpolation.C (hunk 2 above)
Comment (jiangwen84) [reply framework/src/utils/LinearInterpolation.C]: > The style issue can be addressed later if you need this PR now

I am afraid I might forget later so I am pushing a new PR to fix it now.

**Review by dschwen — APPROVED** [review]

**Pull request merged**